src/libstd/sys/windows/stdio.rs

```diff
@@ -1,140 +1,226 @@
     if handle == c::INVALID_HANDLE_VALUE {
         Err(io::Error::last_os_error())
     } else if handle.is_null() {
         Err(io::Error::from_raw_os_error(c::ERROR_INVALID_HANDLE as i32))
     } else {
-        let mut out = 0;
-        match unsafe { c::GetConsoleMode(handle, &mut out) } {
-            0 => Ok(Output::Pipe(handle)),
-            _ => Ok(Output::Console(handle)),
-        }
+        Ok(handle)
     }
 }
 
-fn write(handle: c::DWORD, data: &[u8]) -> io::Result<usize> {
-    let handle = match get(handle)? {
-        Output::Console(c) => c,
-        Output::Pipe(p) => {
-            let handle = Handle::new(p);
-            let ret = handle.write(data);
-            handle.into_raw();
-            return ret
-        }
-    };
+fn is_console(handle: c::HANDLE) -> bool {
+    // `GetConsoleMode` will return false (0) if this is a pipe (we don't care about the reported
+    // mode). This will only detect Windows Console, not other terminals connected to a pipe like
+    // MSYS. Which is exactly what we need, as only Windows Console needs a conversion to UTF-16.
+    let mut mode = 0;
+    unsafe { c::GetConsoleMode(handle, &mut mode) != 0 }
+}
 
-    // As with stdin on windows, stdout often can't handle writes of large
-    // sizes. For an example, see #14940. For this reason, don't try to
-    // write the entire output buffer on windows.
-    //
-    // For some other references, it appears that this problem has been
-    // encountered by others [1] [2]. We choose the number 8K just because
-    // libuv does the same.
+fn write(handle_id: c::DWORD, data: &[u8]) -> io::Result<usize> {
+    let handle = get_handle(handle_id)?;
+    if !is_console(handle) {
+        let handle = Handle::new(handle);
+        let ret = handle.write(data);
+        handle.into_raw(); // Don't close the handle
+        return ret;
+    }
+
+    // As the console is meant for presenting text, we assume bytes of `data` come from a string
+    // and are encoded as UTF-8, which needs to be encoded as UTF-16.
     //
-    // [1]: https://tahoe-lafs.org/trac/tahoe-lafs/ticket/1232
-    // [2]: http://www.mail-archive.com/log4net-dev@logging.apache.org/msg00661.html
-    const OUT_MAX: usize = 8192;
-    let len = cmp::min(data.len(), OUT_MAX);
+    // If the data is not valid UTF-8 we write out as many bytes as are valid.
+    // Only when there are no valid bytes (which will happen on the next call), return an error.
+    let len = cmp::min(data.len(), MAX_BUFFER_SIZE);
     let utf8 = match str::from_utf8(&data[..len]) {
         Ok(s) => s,
-        Err(ref e) if e.valid_up_to() == 0 => return Err(invalid_encoding()),
+        Err(ref e) if e.valid_up_to() == 0 => {
+            return Err(io::Error::new(io::ErrorKind::InvalidData,
+                "Windows stdio in console mode does not support non-UTF-8 byte sequences; \
+                see https://github.com/rust-lang/rust/issues/23344"))
```

Comment:
I think the issue reference may no longer be relevant? In a case like this I'm not really sure what's possible for us to do :(

Reply:
Well I suppose an alternate strategy is that we skip the byte and start spraying utf-8 replacement characters everywhere...

Reply:
> I think the issue reference may no longer be relevant?

I didn't want to touch the error message too much, it was only added last week in https://github.com/rust-lang/rust/pull/58136. But I don't fully agree with its mention of https://github.com/rust-lang/rust/issues/23344. cc @abonander

Reply:
While the issue has some contextual information I think it's likely to get dated over time and I don't think it's appropriate for this situation. Linking to an issue makes sense if it's a bug in libstd for example, but this isn't a bug in libstd it's just a 'fact of life' about the implementation.

Reply:
The error message before was uselessly cryptic and didn't explain the issue at all: https://github.com/abonander/multipart/issues/106

#23344 was the most relevant reference I could find for context. I don't care if that's dropped as long as the error message continues to explain the specific situation and doesn't read like a generic UTF-8 decode error like it did before.